libcudacxx/include/cuda/std/__complex/roots.h

```diff
@@ -24,38 +24,189 @@
 
+// Specialization of 1/sqrt(x) to use rsqrt on device when available.
+// Curiously, for its use here in csqrt, since the device fp32 rsqrtf
+// maps directly to device asm, while the fp64 rsqrt maps to an llvm-ir
+// intrinsic, with the compiler optimizations 1/sqrt(x) generates faster
+// code than rsqrt(x) for fp64. While rqsrtf is better for fp32.
+// So only specialize for fp32.
+template <class _Tp>
+[[nodiscard]] _CCCL_API inline _Tp __internal_rsqrt(_Tp __x)
+{
+  return _Tp(1.0) / _CUDA_VSTD::sqrt(__x);
+}
+
+template <>
+_CCCL_API inline float __internal_rsqrt<float>(float __x)
+{
+  // Use device optimized rsqrt when better.
+  float __sqrt_x;
+
+  NV_IF_ELSE_TARGET(NV_IS_DEVICE, (__sqrt_x = ::rsqrtf(__x);), (__sqrt_x = 1.0f / (_CUDA_VSTD::sqrtf(__x));))
+  return __sqrt_x;
+}
+
 // sqrt
 
 template <class _Tp>
 [[nodiscard]] _CCCL_API inline complex<_Tp> sqrt(const complex<_Tp>& __x)
 {
-  if (_CUDA_VSTD::isinf(__x.imag()))
+  // Uint of the same size as our fp type.
+  // Shouldn't need make_unsigned, but just in case:
+  using __uint_t = make_unsigned_t<__fp_storage_of_t<_Tp>>;
+
+  constexpr int32_t __max_exponent = __fp_exp_max_v<__fp_format_of_v<_Tp>>;
+  constexpr int32_t __mant_nbits   = __fp_mant_nbits_v<__fp_format_of_v<_Tp>>;
+  constexpr int32_t __exp_bias     = __fp_exp_bias_v<__fp_format_of_v<_Tp>>;
+
+  const _Tp __re = __x.real();
+  const _Tp __im = __x.imag();
+
+  if (_CUDA_VSTD::isinf(__im))
   {
-    return complex<_Tp>(numeric_limits<_Tp>::infinity(), __x.imag());
+    return complex<_Tp>(_CUDA_VSTD::numeric_limits<_Tp>::infinity(), __im);
   }
-  if (_CUDA_VSTD::isinf(__x.real()))
+
+  if ((__re == _Tp(0.0)) && (__im == _Tp(0.0f)))
   {
-    if (__x.real() > _Tp(0))
-    {
-      return complex<_Tp>(__x.real(),
-                          _CUDA_VSTD::isnan(__x.imag()) ? __x.imag() : _CUDA_VSTD::copysign(_Tp(0), __x.imag()));
-    }
-    return complex<_Tp>(_CUDA_VSTD::isnan(__x.imag()) ? __x.imag() : _Tp(0),
-                        _CUDA_VSTD::copysign(__x.real(), __x.imag()));
+    return complex<_Tp>(_Tp(0.0), __im);
   }
-  return _CUDA_VSTD::polar(_CUDA_VSTD::sqrt(_CUDA_VSTD::abs(__x)), _CUDA_VSTD::arg(__x) / _Tp(2));
+
+  // pre-check to see if we over/underflow:
+  _Tp __x_abs_sq = _CUDA_VSTD::fma(__re, __re, __im * __im);
+
+  // Get some bounds where __re +- |__x| won't overflow.
+  // Doesn't need to be too exact, enough to cover extremal cases.
+  // overflow bound = sqrt(MAX_FLOAT / 2)
+  // underflow bound similar, but tweaked to allow for normalizing denormal calculation.
+  constexpr __uint_t __overflow_bound_exp =
+    __uint_t(((__max_exponent - 1) >> 1) + __exp_bias) << __uint_t(__mant_nbits);
+  constexpr __uint_t __underflow_bound_exp =
+    __uint_t(((-__max_exponent + __mant_nbits) >> 1) + __exp_bias) << __uint_t(__mant_nbits);
+
+  _Tp __overflow_bound  = _CUDA_VSTD::bit_cast<_Tp>(__overflow_bound_exp);
+  _Tp __underflow_bound = _CUDA_VSTD::bit_cast<_Tp>(__underflow_bound_exp);
```

Comment:
```suggestion
  _Tp __overflow_bound  = _CUDA_VSTD::__fp_from_storage<_Tp>(__overflow_bound_exp);
  _Tp __underflow_bound = _CUDA_VSTD::__fp_from_storage<_Tp>(__underflow_bound_exp);
```
Please, prefer `__fp_from_storage` over `bit_cast`

Reply:
The signature is:
```cpp
constexpr _Tp __fp_from_storage(__fp_storage_of_t<_Tp> __v);
```
The idea is to provide uniform API for all FP types. `bit_cast` would be fine, but it's not `constexpr` for older GCC versions. This funciton allows us to construct the types such as `__half` and `__nv_bfloat16` in constexpr context without the need of `bit_cast`..